llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeAttention.cpp

```diff
@@ -245,24 +266,67 @@ createAttentionBody(Value keySlice, Value valueSlice, Value querySlice,
       updateAndScale(maxSlice, newMax, sumSlice, loc, builder, ops);
   Value newSum = computeRowwiseReduction<arith::AddFOp>(
       partialSoftmax, scaledOldSum, loc, builder, ops);
+  Value inverseNewSum = computeReciprocal(newSum, loc, builder, ops);
   Value softmax =
-      scalePartialSoftmax(partialSoftmax, newSum, loc, builder, ops);
+      scalePartialSoftmax(partialSoftmax, inverseNewSum, loc, builder, ops);
+  if (elementType.isF16()) {
+    Value empty =
+        builder.create<tensor::EmptyOp>(loc, resultShape, builder.getF16Type());
+    softmax = truncateToF16(softmax, empty, ops, builder, loc);
+  }
 
   // Update accumulator
-  Value empty = builder.create<tensor::EmptyOp>(
-      loc, SmallVector<OpFoldResult>{sequenceTileLength, headDimension},
-      elementType);
-  Value scaledAcc = scaleAccumulator(outputSlice, scaledOldSum, newSum, empty,
+  Value scaledAcc = scaleAccumulator(outputSlice, scaledOldSum, inverseNewSum,
                                      loc, builder, ops);
 
   // Compute matmul(softmax, v)
   auto matmulOp = builder.create<linalg::MatmulOp>(
-      loc, outputSlice.getType(), ValueRange{softmax, valueSlice}, scaledAcc);
+      loc, scaledAcc.getType(), ValueRange{softmax, valueSlice}, scaledAcc);
   ops.push_back(matmulOp);
   Value result = matmulOp.getResult(0);
   return std::make_tuple(result, newMax, newSum);
 }
 
+static Value extractOrInsertOutputSlice(Value src, Value dst,
+                                        ArrayRef<int64_t> queryShape,
+                                        OpFoldResult sequenceTileLength,
+                                        OpFoldResult headDimension,
+                                        Location loc, OpBuilder &builder) {
+  auto one = builder.getIndexAttr(1);
+  auto zero = builder.getIndexAttr(0);
+  SmallVector<OpFoldResult> strides(3, one);
+  SmallVector<OpFoldResult> sizes{one, sequenceTileLength, headDimension};
```

Comment:
Prefer using assignment-like with `=` when initializing with the intended literal value. It let us have uniform style, and avoid bugs like below:

```
std::vector<std::string> strings{2}; // A vector of two empty strings.
std::vector<int> ints{2};            // A vector containing only the integer 2.
```

Reply:
Sure, will remove the braces initializer. Thanks!